File: canonical/python/src/yx/transport/udp_socket.py
"""UDP Socket Configuration for YX Protocol."""

import socket
from typing import Tuple
from .packet_builder import PacketBuilder


class UDPSocket:
    """Configure and manage UDP socket for YX protocol."""

    def __init__(self, port: int = 50000):
        """
        Initialize UDP socket.

        Args:
            port: Port to bind to (default: 50000)
        """
        self.port = port
        self.socket: socket.socket = None

    def create_socket(self) -> socket.socket:
        """
        Create and configure UDP socket.

        Returns:
            socket.socket: Configured UDP socket

        Raises:
            OSError: If socket creation fails
        """
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

        # Allow address reuse
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

        # Allow port reuse (multiple listeners on same port)
        try:
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        except AttributeError:
            # SO_REUSEPORT not available on Windows
            pass

        # Enable broadcast
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

        self.socket = sock
        return sock

    def bind(self):
        """Bind socket to all interfaces."""
        if self.socket is None:
            self.create_socket()
        self.socket.bind(('0.0.0.0', self.port))

    def close(self):
        """Close socket."""
        if self.socket:
            self.socket.close()
            self.socket = None

    def send_packet(self, guid: bytes, payload: bytes, key: bytes, host: str = '255.255.255.255', port: int = 50000):
        """Send YX packet via UDP."""
        if self.socket is None:
            self.create_socket()

        data = PacketBuilder.build_and_serialize(guid, payload, key)
        self.socket.sendto(data, (host, port))

    def receive_packet(self, key: bytes, buffer_size: int = 65507) -> Tuple[bytes, bytes, Tuple[str, int]]:
        """
        Receive and parse YX packet.

        Returns:
            (guid, payload, addr) tuple or raises exception
        """
        if self.socket is None:
            raise RuntimeError("Socket not created")

        data, addr = self.socket.recvfrom(buffer_size)
        packet = PacketBuilder.parse_and_validate(data, key)

        if packet is None:
            raise ValueError("Invalid packet received")

        return packet.guid, packet.payload, addr

File: swift/builds/swift-impl/Sources/YXProtocol/Transport/UDPSocket.swift
import Foundation

// Implements: protocol/specs/technical/yx-protocol-spec.md § UDP Transport

public class UDPSocket {
    public let port: UInt16
    private var sockfd: Int32 = -1
    private var timeoutSeconds: Double = 5.0

    public init(port: UInt16 = 50000) {
        self.port = port
    }

    /// Bind socket to the port for receiving.
    public func bind() throws {
        sockfd = socket(AF_INET, SOCK_DGRAM, 0)
        guard sockfd >= 0 else { throw POSIXError(.ENOTSUP) }

        var opt: Int32 = 1
        setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &opt, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = port.bigEndian
        addr.sin_addr.s_addr = INADDR_ANY.bigEndian

        let bindResult = withUnsafePointer(to: addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Foundation.bind(sockfd, sockPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bindResult == 0 else { throw POSIXError(.EADDRINUSE) }
    }

    /// Set receive timeout in seconds.
    public func setTimeout(_ seconds: Double) {
        timeoutSeconds = seconds
        var tv = timeval()
        tv.tv_sec = Int(seconds)
        tv.tv_usec = Int32((seconds - Double(Int(seconds))) * 1_000_000)
        setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))
    }

    /// Receive one validated YX packet. Returns (guid, payload, senderAddr) or nil on timeout.
    public func receivePacket(key: Data) throws -> (Data, Data, String)? {
        let symKey = try Data.symmetricKey(from: key)
        var buffer = [UInt8](repeating: 0, count: 65536)
        var senderAddr = sockaddr_in()
        var senderAddrLen = socklen_t(MemoryLayout<sockaddr_in>.size)

        let n = withUnsafeMutablePointer(to: &senderAddr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                recvfrom(sockfd, &buffer, buffer.count, 0, sockPtr, &senderAddrLen)
            }
        }

        if n <= 0 { return nil } // timeout

        let data = Data(buffer[0..<n])
        guard let packet = PacketBuilder.parseAndValidate(data, key: symKey) else {
            return nil // invalid HMAC
        }

        let addrStr = String(cString: inet_ntoa(senderAddr.sin_addr))
        return (packet.guid, packet.payload, addrStr)
    }

    /// Send data via UDP.
    public func sendPacket(data: Data, host: String = "127.0.0.1", port: UInt16) throws {
        let tempSock = socket(AF_INET, SOCK_DGRAM, 0)
        guard tempSock >= 0 else { throw POSIXError(.ENOTSUP) }
        defer { Darwin.close(tempSock) }

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = port.bigEndian
        addr.sin_addr.s_addr = inet_addr(host)

        try data.withUnsafeBytes { ptr in
            let sent = withUnsafePointer(to: addr) { addrPtr in
                addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockAddrPtr in
                    sendto(tempSock, ptr.baseAddress, data.count, 0, sockAddrPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
            if sent < 0 { throw POSIXError(.ENOTSUP) }
        }
    }

    public func close() {
        if sockfd >= 0 {
            Darwin.close(sockfd)
            sockfd = -1
        }
    }

    deinit {
        close()
    }
}
